### PR TITLE
feat(cli): add v2 deployment script

### PR DESCRIPTION
**Pull request by alvrs**
Deployment script for v2 contracts (World, tables, systems).
- Requires a `PRIVATE_KEY` .env variable
- Uses rpc url from `foundry.toml` (profile can be switched with the `--profile` param)
- Uses `mud.config.mts`
- A compatible example repo can be found here: https://github.com/latticexyz/v2sandbox/tree/alvrs/deploy-v2

### CODE REVIEW

File: packages/cli/src/commands/deploy-v2.ts
Comment (alvrs) [inline packages/cli/src/commands/deploy-v2.ts:86]: I wonder if most of the logic in this function should be moved into the `zod` validators. I put it here for now because it depends on running `forge build` and then grabbing all `*System.sol` contracts from the file system, which felt like something that doesn't really fit into `zod`. Thoughts @holic @dk1a?

Comment (dk1a) [reply packages/cli/src/commands/deploy-v2.ts:86]: I agree it doesn't fit into the world's schema.
But you could add a separate function that accepts `loadWorldConfig`'s output and the results that depend on `forge`. The main benefit there is not so much a zod schema though, but having that function separated from cli

Comment (dk1a) [reply packages/cli/src/commands/deploy-v2.ts:86]: Kinda like how my `tablegen` command does just some filesystem writing, and `renderTablesFromConfig` is a separate exported function you can use on its own

Comment (dk1a) [reply packages/cli/src/commands/deploy-v2.ts:86]: `renderTablesFromConfig` also doesn't really do any validation (you would need some of course), but I think you don't need to validate world config again. You can just export the parseConfig function as well

Comment (alvrs) [reply packages/cli/src/commands/deploy-v2.ts:86]: Another thought I just had is we could pass `existingContractNames` as array into `loadWorldConfig`, so it can be used for validation inside `zod` (checking that the system overrides and excludes only refer to existing contracts). And we could probably use something like `zod`'s `transform` to resolve `accessList` into `accessListSystems` and `accessListAddresses`. Downside is `forge build` would happen before validating the config, but it should be really quick if no contracts change in between the config changes.

File: packages/cli/src/commands/deploy-v2.ts
Comment (holic) [inline packages/cli/src/commands/deploy-v2.ts:46]: should this be `${outDir}/**/*System.sol` to also capture system files in subdirectories?

File: packages/cli/src/commands/deploy-v2.ts
Comment (holic) [reply packages/cli/src/commands/deploy-v2.ts:86]: I defer to @dk1a, who now has more zod experience than me haha

Comment (dk1a) [reply packages/cli/src/commands/deploy-v2.ts:86]: > Another thought I just had is we could pass existingContractNames as array into loadWorldConfig, so it can be used for validation inside zod

I haven't really explored your code yet, but it doesn't seem like `existingContractNames` should be a part of the config loader (unless I'm missing something and it's a config parameter).

To summarize: I'm suggesting another little zod schema and a separate method that accepts config+forge stuff, and validates/transforms those lists. And it can assume world config is otherwise validated already

File: packages/cli/src/commands/deploy-v2.ts
Comment (alvrs) [reply packages/cli/src/commands/deploy-v2.ts:46]: It seems like forge flattens the files in the output directory - in my test project my system is at `src/systems/IncrementSystem.sol`, the output is at `out/IncrementSystem.sol`

Comment (holic) [reply packages/cli/src/commands/deploy-v2.ts]: oh look, I've already reported this 🙈  https://github.com/foundry-rs/foundry/issues/2079

**Comment by holic** [comment]
is [alvrs/sync-store-v2-hack](https://github.com/latticexyz/mud/tree/alvrs/sync-store-v2-hack) the right base branch?

File: packages/cli/src/config/loadWorldConfig.ts
Comment (holic) [inline packages/cli/src/config/loadWorldConfig.ts:41]: What about just `systems`? 

and `exclude` could be an option, e.g.
```
systems: {
  SomeSystemToOverride: {...},
  SomeSystemToExclude: { exclude: true }
}
```

**Review by holic — COMMENTED** [review]
Are there any artifacts from this deploy process where we know what our system calls are going to be? Like will we know how to make a system call via the world?

File: packages/cli/src/utils/schema.ts
Comment (dk1a) [reply packages/cli/src/utils/schema.ts:10]: solidity one uses Bytes extensively, and I wouldn't wanna duplicate that in schema-type.
I think it could be useful to make a `solidity-utils` first, put `Bytes`, `Memory` and maybe some others in there, and then use that in `schema-type` to do schema encoding

File: packages/cli/src/config/loadWorldConfig.ts
Comment (dk1a) [inline packages/cli/src/config/loadWorldConfig.ts:171]: Should probably mention the default

File: packages/cli/src/utils/foundry.ts
Comment (dk1a) [inline packages/cli/src/utils/foundry.ts:4]: this seems like a pretty generic wrapper, maybe it should be in its own file, like a replacement for execLog?

**Review by dk1a — COMMENTED** [review]
lgtm, just some nits
Using ethers doesn't look particularly easier than LibDeploy, and certainly has its own share of problems. But I think I do prefer to not have a temporary script file if we can avoid it

File: packages/cli/src/utils/foundry.ts
Comment (holic) [reply packages/cli/src/utils/foundry.ts:4]: "safe" is also a misleading term here (since executing arbitrary commands is ~rarely safe)

It looks like `execa` exports `execaSync` which seems to do what you're doing here, but maybe more robust:
https://github.com/sindresorhus/execa#execasyncfile-arguments-options

**Comment by alvrs** [comment]
> is [alvrs/sync-store-v2-hack](https://github.com/latticexyz/mud/tree/alvrs/sync-store-v2-hack) the right base branch?

No, need to rebase on main before merging. But it shows the correct diffs.

File: packages/cli/src/utils/deploy-v2.ts
Comment (alvrs) [reply packages/cli/src/utils/deploy-v2.ts:206]: The reason for manual nonce management here is so we can send the transactions without waiting for previous transactions. Without this ethers errored due to duplicate nonces. Unless ethers has an option for this use case that I'm not aware of?

File: packages/cli/src/config/loadWorldConfig.ts
Comment (alvrs) [reply packages/cli/src/config/loadWorldConfig.ts:41]: imo `overrideSystems` makes it more obvious that systems not mentioned here are included by default. If it was just called `systems`, I would assume only systems mentioned here to be deployed. In `resovleWorldConfig` we then use the `overrideSystems` and default systems to build the full `systems` key, so the resulting config object only includes the `systems` key (not `overrideSystems` or `excludeSystems`)

**Comment by alvrs** [comment]
> Using ethers doesn't look particularly easier than LibDeploy, and certainly has its own share of problems.

While working on this deploy script I realized that a bunch of things we can do in js would be much harder to realize in a codegenerated solidity script (like registering nested routes, or more advanced features that will come later like "only deploy a contract if it doesn't yet exist on-chain"). But I agree the most important factor is the advantage of not having to generate a temporary file - `LibDeploy.sol not found` has been a very common error for users.

File: packages/cli/src/config/loadWorldConfig.ts
Comment (alvrs) [reply packages/cli/src/config/loadWorldConfig.ts:10]: This went through a bunch of iterations but the final one is very basic and probably doesn't require the comment 👍

File: packages/cli/src/utils/foundry.ts
Comment (alvrs) [reply packages/cli/src/utils/foundry.ts:4]: Yeah not sure why this is called `safeExec`, should definitely be something like `execLog`, and agree with moving it to a general location.

Re `execaSync` - the one here is more of a helper to pipe the `stdout` and throw `stderr` in the main process (plus some default logging), while `execaSync` is just the non-async version of `execa`, so we'd still need the wrapper for setting the `stdio` options I think. Also being async is intended here, because it allows us to execute multiple commands without waiting for the result of the previous one.

**Comment by alvrs** [comment]
> Are there any artifacts from this deploy process where we know what our system calls are going to be? Like will we know how to make a system call via the world?

I think we can create types and utils (generated or even dynamic) for system calls from config + typechain/abitype. That's what I'll work on next, but gonna put it in a follow up PR.

File: packages/cli/src/utils/deploy-v2.ts
Comment (alvrs) [reply packages/cli/src/utils/deploy-v2.ts:152]: Unfortunately foundry doesn't seem to pick up the `PRIVATE_KEY` environment variable by default to replace the `--private-key` argument, but only provides a `vm` method to load it in foundry scripts. But since we use ethers to deploy / call contracts it doesn't really matter. I've replaced our `privateKey` cli param with loading the `PRIVATE_KEY` environment variable and never logging it in clear text.

Comment (alvrs) [reply packages/cli/src/utils/deploy-v2.ts:152]: And also moved to using the rpc url from `foundry.toml` (and adding a `profile` param to our cli, which acts both as selector for the foundry profile, and we can use it to have default logic based on the profile, eg automatically funding the wallet when deploying to the lattice testnet).

Comment (alvrs) [reply packages/cli/src/utils/deploy-v2.ts:152]: Looks like `FOUNDRY_PROFILE` is ignored in `cast block-number`, maybe also in other cast commands? Reported here: https://github.com/foundry-rs/foundry/issues/4474, will still load the rpc url from the foundry.toml but also pass as an explicit argument to cast for now.

File: packages/cli/src/utils/deploy-v2.ts
Comment (holic) [reply packages/cli/src/utils/deploy-v2.ts:212]: this is what I was referring to: https://docs.ethers.org/v5/api/experimental/#experimental-noncemanager

File: packages/cli/src/utils/deploy-v2.ts
Comment (alvrs) [reply packages/cli/src/utils/deploy-v2.ts:54]: It does break if someone calls their system `World` with the intention of deploying it as a system, but it does allow for users to define their own contract called `World` to use as the world contract with the default config (it's also possible for them to use a contract not called `World` as World by specifying the `worldContractName` config param).

I can add a check for preventing `World` as key in `overrideSystems` to make it more explicit that `World` is not a valid system name though.

**Comment by alvrs** [comment]
Implemented feedback and added a couple cleanups and small features (like auto-retrying transactions if the error is recoverable). Ready for another look. (Commits since your last review: https://github.com/latticexyz/mud/pull/450/files/eabfc00c6a489a2ecb6a7fd4eac1e559fbca80fb..9250d5c466f0ccbdf52c9471de090b8a2a05c899)

File: packages/cli/src/utils/deploy-v2.ts
Comment (alvrs) [reply packages/cli/src/utils/deploy-v2.ts]: Oh I see, wasn't aware of that! I'd still keep the manual nonce handling for now since it's pretty simple and might help with porting to viem.

**Review by dk1a — APPROVED** [review]
lgtm
schema encoder may be better off in its own file, but that can be done together with the other refactoring (after moving solidity Schema)

File: packages/cli/src/commands/deploy-v2.ts
Comment (alvrs) [reply packages/cli/src/commands/deploy-v2.ts:33]: Added this!

File: packages/cli/src/utils/foundry.ts
Comment (alvrs) [reply packages/cli/src/utils/foundry.ts]: all environment variables are part of `process.env` (because we use `dotenv` in `mud.ts`), and everything that is part of `process.env` is automatically included in `execa`'s `env`. This means we actually wouldn't need a `--profile` flag in our cli, but I think it's still useful to explicitly know the profile in our cli, so we can 1. log it to the user and 2. add more advanced features later, like auto-funding accounts via faucets if a certain profile is used (which could arguably also be done via a check on the rpc url, but I still think being able to set the profile via a cli param in addition to falling back to the env variable is good ux).

**Review by holic — APPROVED** [review]

File: packages/cli/src/utils/foundry.ts
Comment (alvrs) [reply packages/cli/src/utils/foundry.ts]: Personally I find the ux of using a flag a bit nicer than writing `FOUNDRY_PROFILE=profile` in front of the comment, but this is not a strong opinion. The other thing I was envisioning was adding default options based on the profile arg itself, instead of requiring them in the foundry config. (Eg `--profile lattice-testnet` -> default values for rpc, auto-funding the deployer wallet from the faucet, etc). Using the `foundry.toml` / `FOUNDRY_PROFILE` env variable feels like an unnecessary indirection here. We could separate that use-case from the "foundry profile" arg though. In any case I propose doing this in a follow up.

Comment (alvrs) [reply packages/cli/src/utils/foundry.ts]: Oh one other thing: these are utils, where a consumer might want to explicitly execute one of them with a profile other than the one found in `FOUNDRY_PROFILE`. Defaulting to the `FOUNDRY_PROFILE` happens one level higher in the `command` itself.

Comment (holic) [reply packages/cli/src/utils/foundry.ts]: I'm good with all that! Foundry basically has the same: env vars + matching command args.

**Pull request merged**
